### PR TITLE
Editor/Viewer while in a call (Viewer Overlay Call View)

### LINKED ISSUE #8551: Can not control media while editing a document in a call
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

## Steps to reproduce
1. Start a call
2. Share a text file into the chat
3. Edit it with the text app
4. Try to mute, disable your video, etc.

### Expected behaviour
Local media control should be reachable

### Actual behaviour
Controls are behind the text editor

### LINKED ISSUE #9295: Editor/viewer while in a call
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

**Is your feature request related to a problem? Please describe.**
When there is a meeting taking place and someone shares a text file, photo, etc in the chat and you open it, you cannot see the meeting anymore as that area is covered by the file. So in case you are taking meeting notes, it becomes quite difficult to mute/unmute, or tell who is speaking, etc

**Describe the solution you'd like**
Show the meeting in a smaller container on the screen while the editor/viewer is open.
image

Must haves:
- [ ] The video of the speaker (or avatar placeholder in case the video is off)
- [ ] Controls for your own audio and video

Nice-to-haves:
- [ ] A smaller video of yourself in the corner to check your own video (indicated by the grey square in the bottom right of the speaker video)
- [ ] A button to collapse this container down and continue the call in the background (indicated by the chevron-down icon in the top right corner of the speaker video). Much like it is possible to collapse the videos of others in the speaker view.
- [ ] A button to return to the call view (the maximise button in the top left corner of the speaker video)

**Describe alternatives you've considered**
n/a

**Additional context**

Current state in Files, which would require some cleaning up.
image

cc @jancborchardt @AndyScherzinger @nickvergessen 
_(edited to rearrange must-haves and nice-to-haves)_

### PR DESCRIPTION
**Pull request by ShGKme**
### ☑️ Resolves

* Fix #9295
* Fix #8551

### 🖼️ Screenshots

image

### 🚧 Tasks

- [x] Create embedded view.
  - [x] Show auto-pilot view in embedded container
  - [x] Add local audio/video controls on embedded instead of participant's ones
  - [x] Add  draft for maximize button
- [x] Integrate embedded view with file view
  - [x] Switch to embedded view on file view open
  - [x] Close embedded view on file view close
  - [x] Close file view on maximize button 
- [x] Improvement
  - [x] Add minimize/maximize button to embedded view
  - [x] Add local view preview on embedded view
- [x] Design review
- [ ] Fix `ETA: unknown`
  - [ ] Fullscreen mode
  - [ ] With local screensharing
  - [ ] Receiving screensharing
- [x] Everything is fine - refactoring `ETA: 1 day`
  - [x] Move to a separated component
  - [x] Fix tests
  - [x] Make local media controls more reusable
  - [x] Clean-up

### 🏁 Checklist

- [ ] ⛑️ Tests (unit and/or integration) are included or not required
- [x] 📘 API documentation in `docs/` has been updated or is not required
- [ ] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required
- [x] 🔖 Capability is added or not needed

### CODE REVIEW

**Comment by ShGKme** [comment]
## Ready for the first design review

image_2023-04-27_23-03-21

### The differences from the initial issue:

### 1. Buttons styles.
In the original feature request they have transparent background. But in most of similar places in 
both Talk and UI components we have buttons with no background. Background is visible on the hover.

### 2. Mic button has volume indicator same as in the top bar.

### Additional

- Click on the video works as maximize button
- Currently the view is a bit large, 300x400px. I'd say, we can make it resizable on a next step

## Current drawbacks

### 1. Limited call management

To make embedded view simple, there is no other management buttons. A user can switch to a main call view if needed. This is about:
  - Start / stop screensharing
  - Disable/enable other participants
  - General conversation management

### 2. No dynamic aspect ratio - fixed at 4:3

Embedded video works with autopilot. The video source may change. To make it static, aspect ratio is fixed, the video may be cut.

Also no dynamic aspect ratio for a video of yourself. It is very small anyway. Prefer to have a static layout here.

## Next steps

- Check edge cases
- Clean-up

**Comment by nimishavijay** [comment]
Looks really good! Only a few points apart from the ones you mentioned:

- [ ] All the buttons need a hover feedback. This can be the same as the hover feedback for the mute/video buttons when you are in a talk call
- [ ] Creating new file from the chat opens the file without showing the viewer
- [ ] If the view is minimised, the maximise button is not visible anymore because the color is white on a white background. The color on this button should be `--color-main-text` (and shouood adjust automatically from dark mode as well)

image


- [ ] If the sidebar is closed, the viewer should move to the side

image

- [ ] on smaller screens the viewer is pretty big and covers almost half the screen, it can be reduced in width and height by 30%
 
image

It's very nice! :) Super cool!

**Comment by ShGKme** [comment]
> All the buttons need a hover feedback. This can be the same as the hover feedback for the mute/video buttons when you are in a talk call

> If the view is minimised, the maximise button is not visible anymore because the color is white on a white background. The color on this button should be --color-main-text (and shouood adjust automatically from dark mode as well)

Fixed

Light

Dark

> Creating new file from the chat opens the file without showing the viewer

Good catch! Fixed.

> If the sidebar is closed, the viewer should move to the side

Fixed. I was sure, I checked it 😅

> on smaller screens the viewer is pretty big and covers almost half the screen, it can be reduced in width and height by 30%

Now it has floating with a 20% screen width between 250px and 400px:
- 250px on ~1280px screen and smaller
- 400px on ~1920px screen and larger

image

**Review by Antreesy — APPROVED** [review]
Super! Tested on web client.

Couple of remarks:
`Expand` button could inherit the same style as `Scroll to bottom` in chat, to be more visible
image
Round corners for local video are too big. Could we shrink it a bit (selector rule is on screenshot)?
image

File: src/components/CallView/shared/ViewerOverlayCallView.vue [hunk 1]
```diff
@@ -0,0 +1,413 @@
+<!--
+  - @copyright Copyright (c) 2023 Grigorii Shartsev <me@shgk.me>
+  -
+  - @author Grigorii Shartsev <me@shgk.me>
+  -
+  - @license GNU AGPL version 3 or any later version
```
[the remark this answers is not in the document]
Comment (ShGKme) [reply src/components/CallView/shared/ViewerOverlayCallView.vue:6]: I followed this: https://github.com/nextcloud/server/blob/master/contribute/HowToApplyALicense.md

cc @nickvergessen

[the remark this answers is not in the document]
Comment (nickvergessen) [reply src/components/CallView/shared/ViewerOverlayCallView.vue:6]: Should be the machine readable one. If you try it in js files eslint will complain. But the eslint rule seems to skip .vue files.

We should also update the [HowToApplyALicense.md](https://github.com/nextcloud/server/blob/master/contribute/HowToApplyALicense.md) then.

[the remark this answers is not in the document]
Comment (ShGKme) [reply src/components/CallView/shared/ViewerOverlayCallView.vue:6]:
```
$ grep -lr "AGPL-3.0-or-later" src | wc -l
> 114

$ grep -lr "GNU AGPL version 3 or any later version" src | wc -l
> 130
```

Should we update all 130 files?

cc @nickvergessen

**Comment by nimishavijay** [comment]
> Should we keep the button, or remove it, or make it looks like a close button?

I would keep it as you have implemented it. Currently the picture-in-picture view that exists has the same button in the same position. It might be confusing if suddenly it doesn't show up anymore, in case someone has been using this.

image

What do you think? @jancborchardt 

> Expand button could inherit the same style as Scroll to bottom in chat, to be more visible

Totally agreed here!

> Round corners for local video are too big. Could we shrink it a bit 

New screenshot looks great, great catch!

Additionally would suggest we could also move the local video a few pixels down so that it is bottom aligned with the bottom of the audio/video buttons, and a few pixels to the right so that the margin in the bottom and right is equal.

image

**Comment by ShGKme** [comment]
> Additionally would suggest we could also move the local video a few pixels down so that it is bottom aligned with the bottom of the audio/video buttons, and a few pixels to the right so that the margin in the bottom and right is equal.

Good catch, I also don't like then something is "almost but not quite".

Fixed

Small | Large
---|---
image | image

**Comment by nimishavijay** [comment]
Nice! And also move it a few px to the right so that the bottom margin and right margin are equal. currently there is slightly more space to the right than in the bottom :)

**Comment by Antreesy** [comment]
> Nice! And also move it a few px to the right so that the bottom margin and right margin are equal. currently there is slightly more space to the right than in the bottom :)

Something like calc(`overlay border-radius`px - `localVideo border-radius`px) for 'bottom' and 'right' properties, so the frame around will be with constant thickness

**Comment by ShGKme** [comment]
Sorry for a force push, it's was a 1 line typo in the last refactoring commit

**Review by nickvergessen — COMMENTED** [review]
Otherwise solid work!

**Comment by juliushaertl** [comment]
Nice, just tested this with text and collabora to see if there is anything we need to do from the office team side.

One issue I noticed though was that the modal does not play well with other elements that may still be visible from talk:
- Modals like the device settings are hidden behind the viewer modal
- Action menus are cut off as they are layered below the viewer modal

Screenshot 2023-05-04 at 09 57 36

It might be an option to render viewer to a different root element as in https://github.com/nextcloud/server/pull/33347/files#diff-22f78918676c8422a8c807a774435aae85eb2f9feed73d4af629f194092d349bR157

**Comment by nickvergessen** [comment]
That is still tracked in #8709 for a follow up

**Comment by rriemann** [comment]
> IMO, it is fine. When the viewer is open, it is kind of a primary app on the page. You can return to the call view to setup anything and then return back to the file.

I recorded a video for #8709 that demonstrates the issue. Please take extra care for usability. People may not understand easily the concept of primary app and secondary apps and models.


https://user-images.githubusercontent.com/111932/231517426-bdb88451-519a-4090-9ceb-be0498ea6c1f.mp4

**Comment by nickvergessen** [comment]
> I recorded a video for https://github.com/nextcloud/spreed/issues/8709 that demonstrates the issue. Please take extra care for usability. People may not understand easily the concept of primary app and secondary apps and models.

We are totally aware, that's also why #8709 is still open. But at least you can disable your mic/camera now when you joined prior to editing. So yeah, doesn't fix everything. But at at least one of the two cases.

**Pull request merged**